gossipd/routing.c

```diff
@@ -994,3 +1001,86 @@ struct route_hop *get_route(tal_t *ctx, struct routing_state *rstate,
 	/* FIXME: Shadow route! */
 	return hops;
 }
+
+/* Get a singleton array of struct node_connection containing
+ * the channel matching the short_channel_id, which must be
+ * an out connection of the given node. */
+static struct node_connection **
+get_out_node_connection_of(const tal_t *ctx,
+			   struct routing_state *rstate,
+			   const struct node *node,
+			   const struct short_channel_id *short_channel_id)
+{
+	struct node_connection **ncs;
+	int i;
+
+	for (i = 0; i < tal_count(node->out); ++i) {
+		if (structeq(&node->out[i]->short_channel_id, short_channel_id)) {
+			ncs = tal_arr(ctx, struct node_connection*, 1);
+			ncs[0] = node->out[i];
+			return ncs;
+		}
+	}
+
+	return NULL;
+}
+/* Get an array of all channels connected to this node. */
+static struct node_connection **
+get_all_node_connections_of(const tal_t *ctx,
+			    struct routing_state *rstate,
+			    const struct node *node)
+{
+	struct node_connection **ncs;
+
+	u64 num_ins = tal_count(node->in);
+	u64 num_outs = tal_count(node->out);
+	u64 num = num_ins + num_outs;
+
+	ncs = tal_arr(ctx, struct node_connection*, num);
+	memcpy(&ncs[0], node->in, num_ins * sizeof(struct node_connection*));
+	memcpy(&ncs[num_ins], node->out, num_outs * sizeof(struct node_connection*));
+
+	return ncs;
+}
+
+void routing_failure(struct routing_state *rstate,
+		     const struct pubkey *erring_node_pubkey,
+		     const struct short_channel_id *scid,
+		     enum onion_type failcode)
+{
+	const tal_t *tmpctx = tal_tmpctx(rstate);
+	struct node *node;
+	struct node_connection **ncs;
+	int i;
+
+	node = get_node(rstate, erring_node_pubkey);
+
+	/* Get all affected channels. */
+	/* BOLT #4:
+	 *
+	 * - if the NODE bit is set:
+	 *   - SHOULD remove all channels connected with the erring node from
+	 *   consideration.
+	 *
+	 */
+	if ((failcode & NODE) != 0)
+		ncs = get_all_node_connections_of(tmpctx, rstate, node);
+	else
+		ncs = get_out_node_connection_of(tmpctx, rstate, node, scid);
```

Comment:
How about:

```
if (failcode & NODE) {
     for (i = 0; i < tal_count(node->in); i++)
        handle_route_fail(rstate, node->in[i], failcode);
    for (i = 0; i < tal_count(node->out); i++)
        handle_route_fail(rstate, node->out[i], failcode);
} else {
     struct node_connection *nc = get_nc_from_node(node, scid);
     /* Possible, if something else deleted nc since we sent payment */
     if (!nc)
          status_trace("Error %s on node %s scid %s: no such connection any more");
     else
          handle_route_fail(rstate, nc, failcode);
```